src/index/updater/inscription_updater.rs

```diff
@@ -76,66 +92,118 @@ impl<'a, 'db, 'tx> InscriptionUpdater<'a, 'db, 'tx> {
-            )
-          })?
-        }
+      // find existing inscriptions on input aka transfers of inscriptions
+      for (old_satpoint, inscription_id) in
+        Index::inscriptions_on_output(self.satpoint_to_id, tx_in.previous_output)?
+      {
+        let offset = input_value + old_satpoint.offset;
+        floating_inscriptions.push(Flotsam {
+          offset,
+          inscription_id,
+          origin: Origin::Old { old_satpoint },
+        });
+
+        inscribed_offsets.insert(offset);
       }
-    }
 
-    if inscriptions.iter().all(|flotsam| flotsam.offset != 0)
-      && Inscription::from_transaction(tx).is_some()
-    {
-      let flotsam = Flotsam {
-        inscription_id: txid.into(),
-        offset: 0,
-        origin: Origin::New {
-          fee: input_value - tx.output.iter().map(|txout| txout.value).sum::<u64>(),
-        },
+      let offset = input_value;
+
+      // different ways to get the utxo set (input amount)
+      input_value += if let Some(value) = self.value_cache.remove(&tx_in.previous_output) {
+        value
+      } else if let Some(value) = self
+        .outpoint_to_value
+        .remove(&tx_in.previous_output.store())?
+      {
+        value.value()
+      } else {
+        self.value_receiver.blocking_recv().ok_or_else(|| {
+          anyhow!(
+            "failed to get transaction for {}",
+            tx_in.previous_output.txid
+          )
+        })?
       };
 
-      if input_value == 0 {
-        self.update_inscription_location(
-          input_sat_ranges,
-          flotsam,
-          SatPoint {
-            outpoint: unbound_outpoint(),
-            offset: self.unbound_inscriptions,
+      // go through all inscriptions in this input
+      while let Some(inscription) = new_inscriptions.peek() {
+        if inscription.tx_in_index != u32::try_from(input_index).unwrap() {
+          break;
+        }
+
+        // ignore reinscriptions on already inscribed offset (sats)
+        // For now we do not allow reinscriptions (for example iscriptions in same input or on
+        // existing inscribed sat
+        let unbound =
+          inscribed_offsets.contains(&offset) || inscription.tx_in_offset != 0 || input_value == 0;
+
+        let cursed = inscribed_offsets.contains(&offset)
+          || inscription.tx_in_index != 0
+          || inscription.tx_in_offset != 0;
+
+        let inscription_id = InscriptionId {
+          txid,
+          index: id_counter,
+        };
+
+        floating_inscriptions.push(Flotsam {
+          inscription_id,
+          offset,
+          origin: Origin::New {
+            fee: 0,
+            cursed,
+            unbound,
           },
-        )?;
-        self.unbound_inscriptions += 1;
-      } else {
-        inscriptions.push(flotsam);
+        });
+
+        new_inscriptions.next();
+        id_counter += 1;
       }
-    };
+    }
+
+    // still have to normalize over multiple inscriptions per tx and inscription size
```

Comment:
We should at least keep the invariant that the sum of genesis fees corresponds to the fee paid for the transaction - i.e. divide genesis fee by number of inscriptions in input for multiple inscriptions per input. That's not yet a fair allocation, but at least total fees would add up correctly.